cep-repodata-state.md

```diff
@@ -0,0 +1,79 @@
+<table>
+<tr><td> Title </td><td> .state.json files for repodata metadata </td>
+<tr><td> Status </td><td> Draft </td></tr>
+<tr><td> Author(s) </td><td> Wolf Vollprecht &lt;wolf@prefix.dev&gt;</td></tr>
+<tr><td> Created </td><td> Jan 09, 2023</td></tr>
+<tr><td> Updated </td><td> Jan 09, 2023</td></tr>
+<tr><td> Discussion </td><td> https://conda.slack.com/archives/C017F7C0VM3/p1672669131100819 </td></tr>
+<tr><td> Implementation </td><td> https://github.com/mamba-org/mamba/pull/2113 </td></tr>
+</table>
+
+## Abstract
+
+Changing how conda and mamba store metadata about repodata.json downloads.
+
+### Motivation
+
+When conda currently downloads `repodata.json` files from the internet, it stores metadata "inside" the file by adding some JSON keys:
+
+- `_url`: The URL that was requested
+- `_etag`: ETag returned from server
+- `_mod`: Last-Modified header from server
+- `_cache_control`: Cache-Control header from server
+
+These are stored as three string values. 
+
+This is not an ideal approach as it modifies the `repodata.json` file and corrupts e.g. the hash of the file. Also, the repodata files have gotten increasingly large, and parsing these state values can require parsing a large `json` file.
+
+Therefore we propose to store the metadata in a secondary file called `.state.json` file next to the repodata.
+
+Another motivating factor is that for the `jlap` proposal we need to (repeatedly) compute the hash value of the `repodata.json` file -- that only gives correct results straight away when the repodata is stored externally.
+
+Both mamba and conda currently use the same cache folder. If both don't implement the same storage strategy but continue to share the same repodata cache, it would lead to frequent cache busting.
+
+### Specification
+
+```json5
+{
+    // we ensure that state.json and .json files are in sync by storing the file
+    // last modified time in the state file, as well as the file size
+    "file_mtime": {
```

Comment:
Maybe you can store this in an ISO standard DateTime format instead? Similar to the `has_zst.last_checked` date. Or is that not precise enough when dealing with file times?

Reply:
IMO the fractional time `os.stat().st_mtime` would not be a problem otherwise `mtime_ns` would be a decent name.

Reply:
I would have to check how exactly the fractional time is computed though.

Reply:
```
In [51]: datetime.datetime.fromtimestamp(1666095612162394000/1000000000).isoformat()
Out[51]: '2022-10-18T08:20:12.162394'
```

🤔

Reply:
Python's datetime has microsecond resolution. You would always have to convert st_mtime to a datetime, and then compare the two datetimes instead of trying to convert the stored datetime to `.timestamp()` and compare with `st_mtime`.

```
In [39]: datetime.fromtimestamp(Path('Dockerfile').stat().st_mtime, tz=zoneinfo.
    ...: ZoneInfo('UTC')).resolution
Out[39]: datetime.timedelta(microseconds=1)
```

Reply:
I think the fractional time is not great because it makes it appear as if it was nanosecond resolution while it isn't. And then we'd have to match the same behavior in C++ or any other language that wants to share the cache.

It seems more natural to store seconds / nanoseconds since UNIX epoch for the mtime since that is what Python does and we don't need to deal with timezones.

However, I am open to other suggestions as long as we can decide to the same behavior.

Reply:
Don't know about why a . is a problem however a single nanoseconds since epoch number would be tolerable

Reply:
The reason is that we loose precision and a comparison will not be straightforward anymore:

https://docs.python.org/3/library/os.html#os.stat_result.st_ctime_ns

https://ciaranm.wordpress.com/2009/11/15/this-week-in-python-stupidity-os-stat-os-utime-and-sub-second-timestamps/

Reply:
There's something extremely pedantic to be said about timestamps; on the other hand comparing two `st_mtime_ns` is really simple.